src/main/java/org/sagebionetworks/bridge/services/TemplateService.java

```diff
@@ -0,0 +1,254 @@
+
+import org.sagebionetworks.bridge.BridgeUtils;
+import org.sagebionetworks.bridge.dao.CriteriaDao;
+import org.sagebionetworks.bridge.dao.TemplateDao;
+import org.sagebionetworks.bridge.exceptions.BadRequestException;
+import org.sagebionetworks.bridge.exceptions.ConstraintViolationException;
+import org.sagebionetworks.bridge.exceptions.EntityNotFoundException;
+import org.sagebionetworks.bridge.models.Criteria;
+import org.sagebionetworks.bridge.models.CriteriaContext;
+import org.sagebionetworks.bridge.models.CriteriaUtils;
+import org.sagebionetworks.bridge.models.GuidVersionHolder;
+import org.sagebionetworks.bridge.models.PagedResourceList;
+import org.sagebionetworks.bridge.models.ResourceList;
+import org.sagebionetworks.bridge.models.Template;
+import org.sagebionetworks.bridge.models.TemplateType;
+import org.sagebionetworks.bridge.models.studies.Study;
+import org.sagebionetworks.bridge.models.studies.StudyIdentifier;
+import org.sagebionetworks.bridge.validators.TemplateValidator;
+import org.sagebionetworks.bridge.validators.Validate;
+
+@Component
+public class TemplateService {
+
+    private TemplateDao templateDao;
+    
+    private CriteriaDao criteriaDao;
+    
+    private StudyService studyService;
+    
+    private SubstudyService substudyService;
+    
+    @Autowired
+    final void setTemplateDao(TemplateDao templateDao) {
+        this.templateDao = templateDao;
+    }
+    
+    @Autowired
+    final void setCriteriaDao(CriteriaDao criteriaDao) {
+        this.criteriaDao = criteriaDao;
+    }
+    
+    @Autowired
+    final void setStudyService(StudyService studyService) {
+        this.studyService = studyService;
+    }
+    
+    @Autowired
+    final void setSubstudyService(SubstudyService substudyService) {
+        this.substudyService = substudyService;
+    }
+
+    public Template getTemplateForUser(CriteriaContext context, TemplateType type) {
+        checkNotNull(context);
+        checkNotNull(type);
+
+        ResourceList<? extends Template> results = templateDao.getTemplates(
+                context.getStudyIdentifier(), type, null, null, false);
+        for (Template template : results.getItems()) {
+            loadCriteria(template);
+        }
+
+        List<Template> templates = results.getItems().stream().filter(template -> {
+            return CriteriaUtils.matchCriteria(context, template.getCriteria());
+        }).collect(toImmutableList());
+        
+        // The ideal case: one and only one template matches the user's context
+        if (templates.size() == 1) {
+            return templates.get(0);
+        }
+        // If not, fall back to the default specified for this study:
+        Study study = studyService.getStudy(context.getStudyIdentifier());
+        String defaultGuid = study.getDefaultTemplates().get(type.name().toLowerCase());
+        if (defaultGuid != null) {
+            return getTemplate(context.getStudyIdentifier(), defaultGuid);
+        }
```

Comment:
Because matching occurs only if match criteria are provided by the user, it is easy to get false positives (an account with no metadata will match all the records in some cases). So we want developers to unambiguously indicate the fallback. This is also clearer for developers (app configs just return the first one and this has led to implementations that are routinely relying on this behavior, probably without knowing their matching criteria are ambiguous).